graphbolt/src/fused_csc_sampling_graph.cc

```diff
@@ -484,6 +495,309 @@ auto GetTemporalPickFn(
+                                    AT_DISPATCH_INDEX_TYPES(
+                                        node_type_offset_.value().scalar_type(),
+                                        "SubstractNodeTypeOffset", ([&] {
+                                          auto node_type_offset_data =
+                                              node_type_offset_.value()
+                                                  .data_ptr<index_t>();
+                                          subgraph_indices_data_ptr[j] -=
+                                              node_type_offset_data
+                                                  [etype_id_to_src_ntype_id[i]];
+                                        }));
+                                }
+                              }
+                            }));
+
+                        if ((!with_seed_offsets || fanouts.size() == 1) &&
+                            type_per_edge_.has_value()) {
+                          // When hetero graph is sampled as a homo graph, we
+                          // still generate type_per_edge tensor for this
+                          // situation.
+                          AT_DISPATCH_INTEGRAL_TYPES(
+                              subgraph_type_per_edge.value().scalar_type(),
+                              "IndexSelectTypePerEdge", ([&] {
+                                auto subgraph_type_per_edge_data_ptr =
+                                    subgraph_type_per_edge.value()
+                                        .data_ptr<scalar_t>();
+                                auto type_per_edge_data_ptr =
+                                    type_per_edge_.value().data_ptr<scalar_t>();
+                                const auto picked_offset =
+                                    subgraph_indptr_data_ptr[seed_offset];
+                                for (auto j = picked_offset;
+                                     j < picked_offset + picked_number; ++j)
+                                  subgraph_type_per_edge_data_ptr[j] =
+                                      type_per_edge_data_ptr
+                                          [picked_eids_data_ptr[j]];
+                              }));
+                        }
+                      }
+                    }
+                  });
+            }));
+      }));
+
+  torch::optional<torch::Tensor> subgraph_reverse_edge_ids = torch::nullopt;
+  if (return_eids) subgraph_reverse_edge_ids = std::move(picked_eids);
+
+  if (with_seed_offsets && fanouts.size() > 1) {
+    subgraph_indptr -= subgraph_indptr_substract;
+  }
+
+  return c10::make_intrusive<FusedSampledSubgraph>(
+      subgraph_indptr, subgraph_indices, seeds, torch::nullopt,
+      subgraph_reverse_edge_ids, subgraph_type_per_edge, edge_offsets);
+}
+
+template <typename NumPickFn, typename PickFn>
+c10::intrusive_ptr<FusedSampledSubgraph>
+FusedCSCSamplingGraph::TemporalSampleNeighborsImpl(
```

Comment:
Why did we need to copy the whole SampleNeighborsImpl code?

Reply:
If there is a deep reason, we can merge this PR with duplicated code and open an issue to refactor it so that the two code paths are same again.

Reply:
I tried to use template function to merge them but the combination of `if constexpr` and lambda function somehow triggers the bug of MSVC so it couldn't build on windows. I didn't find a way to bypass it so I copied the code. We can try to find a nicer way to merge in another PR.

Reply:
I see. That is unfortunate. Maybe we could try a later MSVC version? We should note this down.

Reply:
I know you don't have a lot of time left but if you can open a PR after this is merged that has the refactored version, we can keep it as reference for later. If you can't do that, it should be okay.

Reply:
Yes that's a good idea